docs/src/lang/variants.md

```diff
@@ -0,0 +1,351 @@
+Apple("red")
+Orange(TRUE)
+```
+
+Variants can wrap records, for when we want to represent compound data with named fields:
+
+```tla
+\* @typeAlias: DRINK =
+\*     Water({ sparkling: Bool })
+\*   | Beer({ malt: Str, strength: Int });
+\*
+\* @type: Bool => DRINK;
+Water(sparkling) == Variant("Water", [ sparkling |-> sparkling ])
+
+\* @type: (Str, Int) => DRINK;
+Beer(malt, strength) == Variant("Beer", [ malt |-> malt, strength |-> strength ])
+
+Once a variant is constructed, it becomes opaque to the type checker, that is,
+the type checker only knows that `Water(TRUE)` and `Beer("Dark", 5)` are both
+of type `DRINK`. This is exactly what we want, in order to combine these values
+in a single set. However, we have lost the ability to access the fields of
+these values. To deconstruct values of a variant type, we have to use other
+operators, presented below.
+
+**Filtering by tag name.** Following the idiomatic use of tagged unions in
+untyped TLA+, we can filter a set of variants:
+
+```tla
+LET Drinks == { Water(TRUE), Water(FALSE), Beer("Radler", 2) } IN
+\E d \in VariantFilter("Beer", Drink):
+    d.strength < 3
+```
+
+We believe that `VariantFilter` is the most commonly used way to partition a
+set of variants. Note that `VariantFilter` transforms a set of variants into a
+set of values (that correspond to the associated tag name).
+
+**Type-safe get.** Sometimes, we do have just a value that does not belong to a
+set, so we cannot use `VariantFilter` directly. In this case, we can use
+`VariantGetOrElse`:
+
+```tla
+LET water == Water(TRUE) IN
+VariantGetOrElse("Beer", water,
+                 [ malt |-> "Non-alcoholic", strength |-> 0])).strength
+```
+
+In the above example, we unpack `water` by using the tag name `"Beer"`.  Since
+`water` is actually tagged with `"Water"`, the operator falls back to the
+default case and returns the record `[ malt |-> "Non-alcoholic", strength |->
+0]`.
+
+**Type-unsafe get.** Sometimes, using `VariantFilter` and `VariantGetOrElse`
+is a nuisance, when we know exactly the value type. In this case, we can bypass
+the type checker and get the value notwithstanding the tag:
+
+```tla
+LET drink == ... IN
+LET nonFree ==
+    IF VariantTag(drink) = "Water"
+    THEN VariantGetUnsafe("Water", drink).sparkling
+    ELSE VariantGetUnsafe("Beer", drink).strength > 0
+IN
+...
+```
+
+In general, you should avoid using `VariantGetUnsafe`, as it is type unsafe :-)
+
+----------------------------------------------------------------------------
+
+## Operators
+
+----------------------------------------------------------------------------
+
+<a name="variantCtor"></a>
+### Variant constructor
+
+**Notation:** `Variant(tagName, associatedValue)`
+
+**LaTeX notation:** same
+
+**Arguments:** Two arguments: the tag name (a string literal) and a value
+(a TLA+ expression).
+
+**Apalache type:** `(Str, a) => tagName(a) | b `, for some types `a` and `b`.
+Note that `tagName` is an identifier in this notation.
```

Comment:
Take some time to explain where `b` comes from in all these examples. A new reader will likely be confused by it.